rs/linux_kernel_command_line/src/lib.rs

```diff
@@ -0,0 +1,354 @@
+                    firstpos.replace(*pos);
+                }
+                res
+            })
+            .map(|(_, x)| x.clone())
+            .collect::<Vec<String>>();
+        firstpos
+    }
+
+    /// Ensure an argument is present in a kernel command line.
+    /// If the argument is in the command line, it is replaced with the supplied
+    /// value in the same position, and all other mentions are removed.
+    /// If the argument is not in the command line, it is appended to the end.
+    pub fn ensure_single_argument(
+        &mut self,
+        argument: &str,
+        value: Option<&str>,
+    ) -> Result<(), UnrepresentableValue> {
+        match self.remove_argument(argument) {
+            Some(pos) => {
+                // match pos >= self.tokenized_arguments.len()
+                self.tokenized_arguments
+                    .insert(pos, Self::format_argument(argument, value)?);
+            }
+            None => {
+                self.add_argument(argument, value)?;
+            }
+        }
+        Ok(())
+    }
+
+    /// Add an argument to the end of a kernel command line.
+    /// This does not replace previous arguments of the same name.
+    pub fn add_argument(
+        &mut self,
+        argument: &str,
+        value: Option<&str>,
+    ) -> Result<(), UnrepresentableValue> {
+        let to_add = Self::format_argument(argument, value)?;
+        self.tokenized_arguments.push(to_add);
+        Ok(())
+    }
+}
+
+impl From<KernelCommandLine> for String {
+    fn from(val: KernelCommandLine) -> Self {
+        val.tokenized_arguments.join(" ")
+    }
+}
+
+impl TryFrom<&str> for KernelCommandLine {
+    type Error = ImproperlyQuotedValue;
+
+    fn try_from(cmdline: &str) -> Result<Self, ImproperlyQuotedValue> {
+        let mut res: Vec<String> = vec![];
+        let mut curr: Vec<char> = vec![];
+        let mut is_quoted = false;
+        for ch in cmdline.chars() {
+            match ch {
+                '"' => {
+                    if is_quoted {
+                        curr.push(ch);
+                        res.push(curr.iter().collect());
+                        curr = vec![];
+                        is_quoted = false;
+                    } else {
+                        curr.push(ch);
+                        is_quoted = true;
+                    }
+                }
+                ' ' => {
+                    if is_quoted {
+                        curr.push(ch);
+                    } else if !curr.is_empty() {
+                        res.push(curr.iter().collect());
+                        curr = vec![];
+                    }
+                }
+                _ => {
+                    curr.push(ch);
+                }
+            }
+        }
+        if !curr.is_empty() {
+            if !is_quoted {
+                res.push(curr.iter().collect());
+            } else {
+                return Err(ImproperlyQuotedValue {
+                    val: curr.iter().collect(),
```

Comment:
Since you're returning here, you can just write `curr` without having to clone it.

Reply:
No, I can't do that.

```
Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
error[E0308]: mismatched types
   --> rs/linux_kernel_command_line/src/lib.rs:163:57
    |
163 |                 return Err(ImproperlyQuotedValue { val: curr });
    |                                                         ^^^^ expected `String`, found `Vec<char>`
    |
    = note: expected struct `String`
               found struct `Vec<char>`
```

Reply:
See my previous comments above `Vec<char>` vs `String`